### PR TITLE
Add method to easily extract parallel errors from ZIO effect

### LINKED ISSUE #853: Add method to easily extract parallel errors from ZIO effect
When effects are combined in parallel, if they fail, then the errors are combined in parallel using `Cause`. However, it's not as easy as it could be to access the parallel errors.

This ticket describes a method that can expose _all_ parallel errors in a single call. The method will be called `parallelErrors and will be added to the `ZIO` trait inside `core` / ... / `ZIO.scala`.

```scala
trait ZIO[-R, +E, +A] { self =>
  ...
  def parallelErrors: ZIO[R, ::[E], A] = ???
  ...
}
```

The method can be implemented by calling `sandbox` on the `self` effect, and then looking at the `Cause` (perhaps using `foldM`) and pulling out all the errors (`E` values). If there is one or more, we can fail the effect returned from `parallelErrors` with `ZIO.fail`; otherwise, if there are no errors but the effect failed for other reasons, we can halt with `ZIO.halt`; and finally, if it did not fail, we can succeed with the success value using `ZIO.succeed`.

### PR DESCRIPTION
**Pull request by softinio**
This resolves #853

### CODE REVIEW

**Comment by softinio** [comment]
@ghostdogpr thanks for your help with this appreciate it. 

Any suggestions on how I should test this?

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala [hunk 1]
```diff
@@ -1102,6 +1102,20 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
+  /**
+   * Exposes all parallel errors in a single call
+   *
+   */
+  final def parallelErrors: ZIO[R, List[E], A] =
+    self.foldCauseM(
+      cause =>
+        cause.failures match {
+          case Nil            => ZIO.halt(cause.asInstanceOf[Cause[Nothing]])
+          case ::(head, tail) => ZIO.fail(::(head, tail))
```
Comment (ghostdogpr) [inline core/shared/src/main/scala/scalaz/zio/ZIO.scala]: `case ::(head, tail) => ZIO.fail(::(head, tail))` can be just `case errors => ZIO.fail(errors)`

**Comment by ghostdogpr** [comment]
This should do the trick to test it:
```scala
f1     <- IO.fail("error1").fork
f2     <- IO.fail("error2").fork
errors <- f1.zip(f2).join.parallelErrors.flip
```

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala [hunk 2]
```diff
@@ -1102,6 +1102,20 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * instance belongs (e.g. `IO.Tags.Succeed`).
    */
   def tag: Int
+
+  /**
+   * Exposes all parallel errors in a single call
+   *
+   */
+  final def parallelErrors: ZIO[R, List[E], A] =
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/ZIO.scala]: I'd change the `List[E]` here to `::[E]`.

Comment (ghostdogpr) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: The problem is that E is invariant in `::` so you get this error: `Error:(398, 13) covariant type E occurs in invariant position in type => scalaz.zio.ZIO[R,::[E],A] of method parallelErrors final def parallelErrors: ZIO[R, ::[E], A]`

Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]:
```scala
final def parallelErrors[E1 >: E]: ZIO[R, ::E1, A] = ...
```

**Comment by jdegoes** [comment]
@softinio Great work on this! 😄

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala (hunk 2 above)
Comment (ghostdogpr) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: 👍

Comment (softinio) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: @jdegoes @ghostdogpr using

 
```
 final def parallelErrors[E1 >: E]: ZIO[R, ::E1, A] = ...
```

I tried this with no success. Same error as

```
final def parallelErrors: ZIO[R, ::E1, A] = ...
```
Using list of course works. 

Any suggestions gentlemen? 

Also whats the benefit of using a cons `::E` over a `List`?

Thanks

Comment (ghostdogpr) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: It is `::[E1]` instead of `::E1` ;)

Comment (softinio) [reply core/shared/src/main/scala/scalaz/zio/ZIO.scala]: you are right I was blind to my mistake. Thanks for that @ghostdogpr

**Comment by jdegoes** [comment]
@softinio Looks great! Circle CI still needs to run on this...

**Comment by softinio** [comment]
@ghostdogpr @jdegoes Trying to add some tests before merging. Started with Pierre suggestion for a test. 

Here is my attempt so far:

https://github.com/zio/zio/pull/917/commits/0ed7a299e7430ad7b3b0258d93abc10955364a55

Getting `missing parameter type` for `errors` in the for comprehension. I don't see why I am getting this error as its clear to me that `errors` is a `::[String]` which is whats expected. Can you be a second pair of eyes and see what I am blind to? 

Good to see metals more inline with scalac than intellij. Metals highlights same error as scalac does, but intellij things its all great and happy :-)

File: core/shared/src/test/scala/zio/ParallelErrorsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+   Returns a list of 2 errors $allFailures
+
+  """
+
+  def allFailures =
+    for {
+      f1     <- IO.fail("error1").fork
+      f2     <- IO.fail("error2").fork
+      errors <- f1.zip(f2).join.parallelErrors.flip
+    } yield errors.length must_=== 2
```
Comment (softinio) [inline core/shared/src/test/scala/zio/ParallelErrorsSpec.scala]: I will improve this to check the cons includes both messages, testing length here whilst wip to start with

**Comment by jdegoes** [comment]
@softinio Can you try providing type parameter to `parallelErrors` and seeing if that fixes the problem?

**Comment by softinio** [comment]
@jdegoes @ghostdogpr This is all ready for your review. Added more tests, all working. Circle passes. Thank you !

**Comment by iravid** [comment]
Quick question: assuming the cause is `Then(Fail(e1), Fail(e2))`, this will expose the sequential failures as parallel failures under this method. Is this intentional?

**Comment by jdegoes** [comment]
We need to fix Cause to disallow that possibility, as it does not make sense

**Comment by iravid** [comment]
Ah because Then should only combine a failure with a defect/interruption?

**Comment by softinio** [comment]
Going to merge this PR, created a separate issue to make the improvement discussed above here: https://github.com/zio/zio/issues/1073

Thanks Gentlemen.

**Pull request merged**